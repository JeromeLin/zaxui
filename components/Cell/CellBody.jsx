
import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';

import Cell from '../Cell';

class CellBody extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: '',
      type: '',
      validate: false,
      dirty: false,
    };
    this.setResult = this.setResult.bind(this);
  }

  setResult(value, validate, opts) {
    const { setResult } = this.props;
    const { type } = opts;

    this.setState({
      value,
      validate,
      type,
      dirty: type !== 'init',
    }, () => {
      setResult && setResult(value, validate, opts);
    });
  }

  helpRender() {
    const { validate, dirty, type } = this.state;
    const { showHelp, help = null } = this.props;

    if (showHelp === 'always') {
      return help;
    }

    if (!validate && dirty) {
      if (showHelp === 'change') {
        return help;
      } else if (showHelp === type) {
        return help;
      }
    }

    return null;
  }

  render() {
    const { children, className, showHelp, setResult, ...otherOpts } = this.props;
    // cell
    const { type, theme, icon, title, description, help, ...others } = otherOpts;
    // validInput
    const { max, min, maxLength, minLength, required, func, pattern, onChange, onBlur, value, ...rest } = others;

    const cls = classnames({
      'ui-cell-form': true,
      [className]: !!className,
    });

    const CellConfig = {
      type,
      theme,
      icon,
      title,
      description,
      ...rest,
    };

    const ValidInputConfig = {
      max,
      min,
      maxLength,
      minLength,
      required,
      func,
      pattern,
      onChange,
      onBlur,
      value,
      ...rest,
    };

    return (
      <Cell className={cls} help={this.helpRender()} {...CellConfig}>
        <Cell.Input setResult={this.setResult} {...ValidInputConfig}>
          {children}
        </Cell.Input>
      </Cell>
    );
  }
}

CellBody.propTypes = {
  type: PropTypes.oneOf(['normal', 'link', 'select']),
  className: PropTypes.string,
  showHelp: PropTypes.oneOf(['always', 'change', 'blur']), // 显示help的时间点，always(一直显示)，change(修改就显示)，blur(失去焦点显示)
};

CellBody.defaultProps = {
  type: 'normal',
  className: null,
  showHelp: 'change',
};

export default CellBody;
