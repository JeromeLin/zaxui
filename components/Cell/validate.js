function _isUndefined(res, func) {
  if (res === undefined) {
    return () => {
      return true;
    };
  }

  return func(res);
}

const min = (res) => {
  return _isUndefined(res, (theRes) => {
    theRes = Number(theRes);

    return (value) => {
      return value >= theRes;
    };
  });
};

const max = (res) => {
  return _isUndefined(res, (theRes) => {
    theRes = Number(theRes);

    return (value) => {
      return value <= theRes;
    };
  });
};

const minLength = (res) => {
  return _isUndefined(res, (theRes) => {
    theRes = Number(theRes);

    return (value) => {
      return value.length >= theRes;
    };
  });
};

const maxLength = (res) => {
  return _isUndefined(res, (theRes) => {
    theRes = Number(theRes);

    return (value) => {
      return value.length <= theRes;
    };
  });
};

const pattern = (res) => {
  return _isUndefined(res, (theRes) => {
    return (value) => {
      return theRes.test && theRes.test(value);
    };
  });
};

const required = (res) => {
  return _isUndefined(res, (theRes) => {
    return (value) => {
      if (!theRes) {
        return true;
      }

      return value !== undefined && value !== '';
    };
  });
};

const func = (res) => {
  return _isUndefined(res, (theRes) => {
    return (value) => {
      if (!theRes) {
        return true;
      }

      return !!theRes(value);
    };
  });
};

const errors = {
  min: ({ title = '', rule = '' } = {}) => `${title} - 不得小于${rule}`,
  max: ({ title = '', rule = '' } = {}) => `${title} - 不得大于${rule}`,
  minLength: ({ title = '', rule = '' } = {}) => `${title} - 长度不得小于${rule}`,
  maxLength: ({ title = '', rule = '' } = {}) => `${title} - 长度不得大于${rule}`,
  pattern: ({ title = '' } = {}) => `${title} - 不符合验证规则`,
  required: ({ title = '' } = {}) => `${title} - 此项必填`,
  func: ({ title = '' } = {}) => `${title} - 验证不通过`,
  default: ({ title = '', value = '' } = {}) => `${title} - ${(value === undefined || value === '') ? '请输入信息' : '已填写'}`,

  // 字段名 当前的值 验证结果 规则约束
  // func: ({ title, value, validate, rule }) => { return '' };
};

const getMsg = (error, opts) => {
  if (!error) {
    return '';
  }

  switch (typeof error) {
    case 'function': {
      return error(opts);
    }
    default: {
      return error;
    }
  }
};

export {
  min,
  max,
  minLength,
  maxLength,
  pattern,
  required,
  func,
  errors,
  getMsg,
};
