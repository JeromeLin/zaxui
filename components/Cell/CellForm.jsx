
import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import Cell from '../Cell';
import ValidForm from '../ValidForm';

class CellForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: '',
      validate: false,
      dirty: false,
    };
    this.setResult = this.setResult.bind(this);
  }

  setResult(value, validate) {
    this.setState({
      value,
      validate,
      dirty: true,
    });
  }

  helpRender() {
    const { validate, dirty } = this.state;
    const { showHelp, help = null } = this.props;

    if (showHelp) {
      return help;
    }

    if (!validate && dirty) {
      return help;
    }

    return null;
  }

  render() {
    const { children, className, showHelp, ...otherOpts } = this.props;
    // cell
    const { type, theme, icon, title, description, help, ...others } = otherOpts;
    // validInput
    const { max, min, maxLength, minLength, required, func, pattern, onChange, onBlur, ...rest } = others;

    const cls = classnames({
      'ui-cell-form': true,
      [className]: !!className,
    });

    const CellConfig = {
      type,
      theme,
      icon,
      title,
      description,
      ...rest,
    };

    const ValidInputConfig = {
      max,
      min,
      maxLength,
      minLength,
      required,
      func,
      pattern,
      onChange,
      onBlur,
      ...rest,
    };

    return (
      <Cell className={cls} help={this.helpRender()} {...CellConfig}>
        <ValidForm.ValidInput setResult={this.setResult} {...ValidInputConfig}>
          {children}
        </ValidForm.ValidInput>
      </Cell>
    );
  }
}

CellForm.propTypes = {
  type: PropTypes.oneOf(['normal', 'link', 'select']),
  className: PropTypes.string,
  showHelp: PropTypes.bool,
};

CellForm.defaultProps = {
  type: 'normal',
  className: null,
  showHelp: false,
};

export default CellForm;
