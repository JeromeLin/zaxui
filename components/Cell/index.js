
import Cell from './Cell';
import Form from './CellForm';
import Validate, { getResult } from './CellValidate';

Cell.Form = Form;
Cell.Validate = Validate;
Cell.getResult = getResult;

export default Cell;
