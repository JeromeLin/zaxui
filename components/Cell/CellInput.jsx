import React, { Component, PropTypes, cloneElement } from 'react';

import * as validateFuncs from './validate';

const typeValidateAttrs = {
  number: ['min', 'max', 'required', 'func'],
  bool: ['required', 'func'],
  string: ['minLength', 'maxLength', 'pattern', 'required', 'func'],
  all: ['min', 'max', 'minLength', 'maxLength', 'pattern', 'required', 'func'],
};

const _getValidate = (attr) => {
  return attr ? {
    validate: false,
    error: attr,
  } : {
    validate: true,
    error: '',
  };
};

class CellInput extends Component {

  constructor(props) {
    super(props);

    this.state = { value: props.value || this._getDefaultValue() };
    this.onChange = this.onChange.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  componentDidMount() {
    this._initResult();
  }

  onChange(e) {
    this._setResult({
      e,
      type: 'change',
      cb: this.props.onChange,
    });
  }

  onBlur(e) {
    this._setResult({
      e,
      type: 'blur',
      cb: this.props.onBlur,
    });
  }

  _initResult() {
    this._setResult({
      type: 'init',
    });
  }

  _setResult({ e, type, cb } = {}) {
    const { name, setResult } = this.props;
    const value = e ? this._getValue(e) : this.state.value;
    const res = this._validate(value);
    const { validate, error } = res;

    const opts = {
      type,
      name,
      error,
    };

    cb && cb(value, validate);
    setResult && setResult(value, validate, opts);
  }

  _getDefaultValue() {
    const { type } = this.props;

    if (type === 'number') {
      return 0;
    }

    if (type === 'bool') {
      return false;
    }

    if (type === 'string') {
      return '';
    }

    if (type === 'all') {
      return '';
    }

    return '';
  }

  _validate(value) {
    const { type = 'all', required } = this.props;
    const attrs = typeValidateAttrs[type] || [];

    if (!required && (value === '' || value === undefined)) {
      return _getValidate();
    }

    for (let v = 0; v < attrs.length; v++) {
      const attr = attrs[v];
      const args = this.props[attr];
      const validateFunc = validateFuncs[attr](args);

      if (!validateFunc(value)) {
        return _getValidate(attr);
      }
    }

    return _getValidate();
  }

  _getValue(e) {
    e = e.target || e;
    const value = e.value === undefined ? e : e.value;

    this.setState({ value });
    return value;
  }

  getChildren() {
    const { onChange, onBlur, props, state } = this;
    const { children } = props;
    const { value } = state;

    return React.Children.map(children, (child) => {
      return cloneElement(child, {
        onChange,
        onBlur,
        value,
      }, child.props.children);
    });
  }

  render() {
    return (
      <span className={this.props.className}>
        {this.getChildren()}
      </span>
    );
  }
}

CellInput.propTypes = {
  max: PropTypes.number,
  min: PropTypes.number,
  maxLength: PropTypes.number,
  minLength: PropTypes.number,
  required: PropTypes.bool,
  func: PropTypes.func,
  // pattern: 正则表达式

  onChange: PropTypes.func,
  onBlur: PropTypes.func,

  setResult: PropTypes.func, // 获取value和验证结果
  resultId: PropTypes.string, // 结果储存在内存的Id
};

CellInput.defaultProps = {
  onChange: () => {},
  onBlur: () => {},
  resultId: '',
};

export default CellInput;
