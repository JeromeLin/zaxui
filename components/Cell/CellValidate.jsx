
import React, { Component, PropTypes, isValidElement } from 'react';
import classnames from 'classnames';

import Cell from '../Cell';
import Icon from '../Icon';
import { errors, getMsg } from './validate';

class CellBody extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: '',
      type: '',
      error: '',
      validate: false,
      dirty: false,
    };
    this.setResult = this.setResult.bind(this);
  }

  setResult(value, validate, opts) {
    const { setResult } = this.props;
    const { type, error } = opts;

    this.setState({
      value,
      validate,
      type,
      error,
      dirty: type !== 'init',
    }, () => {
      setResult && setResult(value, validate, opts);
    });
  }

  getHelpByObj(obj = {}) {
    const error = this.state.error || 'default';
    const curr = Object.assign({}, errors, obj);

    const func = curr[error];
    const res = getMsg(func, this.getOpts());

    return this.getHelp(res);
  }

  getOpts() {
    const { error, validate, value } = this.state;
    const { title } = this.props;

    return {
      title,
      validate,
      value,
      rule: this.props[error],
    };
  }

  getHelp(help) {
    help = help || this.props.help || null;

    let ele = null;
    const { value, validate, error } = this.state;

    if (isValidElement(help)) {
      return help;
    }

    switch (typeof help) {
      case 'string': {
        ele = (
          <span>
            <Icon type="info-round" />
            { help }
          </span>
        );
        break;
      }

      case 'function': {
        ele = help(value, validate, error);
        break;
      }

      case 'obejct': {
        ele = this.getHelpByObj(help);
        break;
      }

      default: {
        ele = this.getHelpByObj();
      }
    }

    return ele;
  }

  helpRender() {
    const { validate, dirty, type } = this.state;
    const { showHelp } = this.props;

    if (showHelp === 'always') {
      return this.getHelp();
    }

    if (!validate && dirty) {
      if (showHelp === 'change') {
        return this.getHelp();
      } else if (showHelp === type) {
        return this.getHelp();
      }
    }

    return null;
  }

  render() {
    const { children, className, showHelp, setResult, ...otherOpts } = this.props;
    // cell
    const { type, theme, icon, title, description, help, ...others } = otherOpts;
    // validInput
    const { max, min, maxLength, minLength, required, func, pattern, onChange, onBlur, value, ...rest } = others;

    const cls = classnames({
      'ui-cell-form': true,
      [className]: !!className,
    });

    const CellConfig = {
      type,
      theme,
      icon,
      title,
      description,
      ...rest,
    };

    const ValidInputConfig = {
      max,
      min,
      maxLength,
      minLength,
      required,
      func,
      pattern,
      onChange,
      onBlur,
      value,
      ...rest,
    };

    return (
      <Cell className={cls} help={this.helpRender()} {...CellConfig}>
        <Cell.Input setResult={this.setResult} {...ValidInputConfig}>
          {children}
        </Cell.Input>
      </Cell>
    );
  }
}

CellBody.propTypes = {
  type: PropTypes.oneOf(['normal', 'link', 'select']),
  className: PropTypes.string,
  showHelp: PropTypes.oneOf(['always', 'change', 'blur', 'hide']), // 显示help的时间点，always(一直显示)，change(修改就显示)，blur(失去焦点显示)，hide(一直隐藏)
};

CellBody.defaultProps = {
  type: 'normal',
  className: null,
  showHelp: 'change',
};

export default CellBody;
