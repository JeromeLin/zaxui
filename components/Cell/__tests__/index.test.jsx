import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Cell from '../index';

describe('Cell', () => {
  it('renders correctly', () => {
    const wrapper = render(<Cell />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
