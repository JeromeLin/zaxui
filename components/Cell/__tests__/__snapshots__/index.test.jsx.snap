// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders correctly 1`] = `
<div
  class="za-cell"
>
  <div
    class="za-cell-inner"
  >
    <div
      class="za-cell-header"
    />
    <div
      class="za-cell-body"
    />
    <div
      class="za-cell-footer"
    />
  </div>
</div>
`;
